src/App/Fossa/Subcommand.hs

```diff
@@ -0,0 +1,50 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module App.Fossa.Subcommand (
+  runSubCommand,
+  EffStack,
+  GetSeverity (..),
+  SubCommand (..),
+) where
+
+import App.Fossa.Config.ConfigFile (ConfigFile)
+import App.Fossa.Config.EnvironmentVars (EnvVars, getEnvVars)
+import Control.Carrier.Diagnostics (DiagnosticsC, context, logWithExit_)
+import Effect.Exec (ExecIOC, runExecIO)
+import Effect.Logger (
+  LoggerC,
+  Severity (SevInfo),
+  withDefaultLogger,
+ )
+import Effect.ReadFS (ReadFSIOC, runReadFSIO)
+import Options.Applicative (InfoMod, Parser)
+
+data SubCommand cli cfg = SubCommand
+  { commandName :: String
+  , commandInfo :: InfoMod (IO ())
+  , parser :: Parser cli
+  , configLoader :: cli -> EffStack (Maybe ConfigFile) -- CLI args can control where we look for config file
+  , optMerge :: Maybe ConfigFile -> EnvVars -> cli -> EffStack cfg
+  , perform :: cfg -> EffStack ()
+  }
+
+type EffStack = ExecIOC (ReadFSIOC (DiagnosticsC (LoggerC IO)))
+
+class GetSeverity a where
+  getSeverity :: a -> Severity
+  getSeverity = const SevInfo
```

Comment:
I think the way we kill `GetSeverity` might be to add a `Logger` operation that allows changing the filter severity, e.g., `withSeverityFilter :: Has Logger sig m => Severity -> m a -> m a`

for the purposes of `runSubCommand`, we would just start with `withDefaultLogger SevInfo`

Reply:
I'd have to check if we would be losing debug info from some of the operations that were moved into the configuration phase, e.g. revision inferrence.  If we aren't missing anything, then this is a good move.

Reply:
So, if we did this, we wouldn't lose any debug info, I've checked thoroughly.

However, I'm not sure this works out, for 2 reasons:

- I'm not sure how we would run `withSeverityFilter` from `runSubCommand` without having some uniform way of accessing the `Severity` anyways. I'd rather have this (admittedly not ideal) typeclass than allow the generic subcommand handler to NOT properly start up the logger.  Notably, the type system can't tell us that we haven't started the "correct" logger effect handler yet.
- How would we actually define `withSeverityFilter`?  
  - Do we just run a new logger underneath?  Are there concerns with running the console handling functions more than once?
  - Do we use the fact that it's a simple effect and somehow create a `local`-like for `SimpleC` that replaces simple effects?

Unless those questions have really simple answers, I'd like to leave this as-is for now.  Happy to leave a TODO and tackle it later, though.

Reply:
oh right -- I forgot `Logger` is a `Simple` effect. that makes things harder. we can just keep this as-is